sourcegraph/learn
# Add local cache for data

### What should this PR do?
Resolves [DEVED-268](https://sourcegraph.atlassian.net/browse/DEVED-268) by removing the page data warnings on `posts` and some tag pages by not passing all of our page data as props, and loading it instead from a local cache.

Some size and speed stats from Chrome:

- `posts`: Branch deploy (33.0 kB 200 ms ), Prod (106 kB 409 ms)
- `tags/sourcegraph`: Branch (24.4 kB 388 ms), Prod (50.5 kB 444 ms)

Also simplifies the logic in our record collections loading methods — rather than loading all files as `MarkdownFile(s)WithUrl` **before** filtering for matches in collection members (which necessitated loading & traversing all records), we instead match the request slug param with collection member strings, and load the relevant collection members that way (so three files instead of all of our files).

### Why are we making this change?
- We want to optimize page performance.
- We want to optimize our loading methods.

### What are the acceptance criteria? 
- No warnings should be throw on `posts` or tag pages, e.g. `Warning: data for page "/posts" is 373 kB, this amount of data can reduce performance.`
- Everything should work exactly as it does on prod for `posts/videos` index pages, and tag index pages, including all record loading functionality.
- All collections should load properly.

### How should this PR be tested?
- Check out the branch & run the app in dev mode with `npm run dev`. This will run a script that creates a local, untracked page data cache under `public/builds`, and starts the app.
- Check the `posts` and `videos` index pages and ensure that the correct records are displayed and loading functionality works as expected.
- Go to some tag pages and ensure that post and video loading works as expected.
- Ensure that in dev mode no warnings are thrown when hitting `posts` or any tag index pages.
- Ensure that all existing collections load properly.

## Pull request process

**Reviewers**:

1. Test functionality using the criteria above.
2. Offer tips for efficiency, feedback on best practices, and possible alternative approaches and things that may not have been considered.
3. For shorter, "quick" PRs, use your best judgement on #​2.
4. Use a collaborative approach and provide resources and/or context where appropriate.
5. Provide screenshots/grabs where appropriate to show findings during review.

**Reviewees**:

1. Prefer incremental and appropriately-scoped changes.
2. Leave a comment on things you want explicit feedback on.
3. Respond clearly to comments and questions.
